replay/models/ucb.py

```diff
@@ -0,0 +1,114 @@
+from typing import Optional
+
+import pandas as pd
+
+from pyspark.sql import DataFrame, Window
+from pyspark.sql import functions as sf
+
+from replay.models.base_rec import Recommender
+
+
+class UCB(Recommender):
+    """
+    Calculates upper confidence bound (UCB) for the confidence interval
+    of true fraction of positive ratings.
+
+    ``relevance`` must be converted to binary 0-1 form.
+
+    >>> import pandas as pd
+    >>> data_frame = pd.DataFrame({"user_idx": [1, 2], "item_idx": [1, 2], "relevance": [1, 1]})
+    >>> from replay.utils import convert2spark
+    >>> data_frame = convert2spark(data_frame)
+    >>> model = UCB()
+    >>> model.fit_predict(data_frame,k=1).toPandas()
+       user_idx  item_idx  relevance
+    0         1         2    2.17741
+    1         2         1    2.17741
+
+    """
+    can_predict_cold_users = True
+    can_predict_cold_items = True
+
+    def _fit(
+            self,
+            log: DataFrame,
+            user_features: Optional[DataFrame] = None,
+            item_features: Optional[DataFrame] = None,
+    ) -> None:
+        vals = log.select("relevance").where(
+            (sf.col("relevance") != 1) & (sf.col("relevance") != 0)
+        )
+        if vals.count() > 0:
+            raise ValueError("Relevance values in log must be 0 or 1")
+
+        items_counts = log.groupby("item_idx").agg(
+            sf.sum("relevance").alias("pos"),
+            sf.count("relevance").alias("total"),
+        )
+
+        full_count = log.count()
+        items_counts = items_counts.withColumn(
+            "relevance",
+            (sf.col("pos") / sf.col("total") + sf.sqrt(sf.log(sf.lit(2 * full_count)) / sf.col("total")))
+        )
+
+        self.item_popularity = items_counts.drop("pos", "total")
+        self.item_popularity.cache()
+
+    @property
+    def _dataframes(self):
+        return {"item_popularity": self.item_popularity}
+
+    def _clear_cache(self):
+        if hasattr(self, "item_popularity"):
+            self.item_popularity.unpersist()
+
+    # pylint: disable=too-many-arguments
+    def _predict(
```

Comment:
Should it be inherited from poprec? 
UPD: I see, you use right join for items filtering. However many lines of code are repeated

Reply:
yes. but what you propose?

Reply:
do the same for poprec and wilson. 
in Wilson relevance is on the interval from 0 to 1, thus you better fillna with 0, which will sout for all three (poprec, wilson, ucb) models

Reply:
ok, 0 relevance here is wrong, but 1 is almost equally wrong, as in most cases an exploration term it is greater than 1. pls, propose a new value

Reply:
I think, good start value is 1+sqrt(c*log(n)), that equivalent to item with one view